### PR TITLE
Restore Static Site features

### PR DESCRIPTION
**Pull request by ansoncfit**
Our old implementation of static sites relied on the browser for propagation (from stops to target grid cells).

This PR now handles much of the propagation (of times and paths) in the R5 worker. 

Features are also added to write static metadata (including the transit network data needed to render transitive.js paths for point-to-point itineraries) and travel time surfaces to S3.

### CODE REVIEW

**Comment by abyrd** [comment]
Great, thanks for posting this work @ansoncfit, I hope to take a look at it today.

File: src/main/java/com/conveyal/r5/analyst/cluster/RegionalTask.java [hunk 1]
```diff
@@ -22,9 +24,17 @@
...
+    /** The grid key on S3 to compute access to. If this is not blank, the default TravelTimeSurfaceTask will be
+     * overridden; returnInVehicleTimes, returnWaitTimes, and returnPaths will be set to false; and the returned results
+     * will be an accessibility value per origin, rather than a grid of travel times from that origin.*/
     public String grid;
 
+    /** An array of grid keys on S3 to compute access to. If this is not blank, the default TravelTimeSurfaceTask will be
+     * overridden; returnInVehicleTimes, returnWaitTimes, and returnPaths will be set to false; and the returned results
+     * will be an accessibility value per origin for each destination grid, rather than a grid of travel times from
+     * that origin.*/
+    public List <String> grids;
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/analyst/cluster/RegionalTask.java]: We can switch to just using a list when the required front-end changes are in place (https://github.com/conveyal/analysis-ui/issues/395)

[the remark this answers is not in the document]
Comment (abyrd) [reply src/main/java/com/conveyal/r5/analyst/cluster/RegionalTask.java]: In my opinion we shouldn't have two fields serving the same purpose. We should only add the list when we're ready to actually use it, or we should add it right away and just change the backend to always send a list of one item until the functionality is added to the frontend.

File: src/main/java/com/conveyal/r5/analyst/TravelTimeComputer.java [hunk 2]
```diff
@@ -280,6 +282,7 @@ public void computeTravelTimes(OutputStream os) throws IOException {
 
                         }
 
+                        // Record the paths used
```
Comment (abyrd) [inline src/main/java/com/conveyal/r5/analyst/TravelTimeComputer.java]: Even the word "paths" is pretty vague unless you're already in the mindset of working on this particular problem. I'd prefer something like "Record the sequence of transit trips and stops used to reach the destination (the paths)."

File: src/main/java/com/conveyal/r5/analyst/Grid.java [hunk 3]
```diff
@@ -428,6 +428,12 @@ public void writeShapefile (String fileName, String fieldName) {
         }
     }
 
+    public boolean hasEqualExtents(Grid comparisonGrid){
+        if (this.zoom != comparisonGrid.zoom || this.west != comparisonGrid.west || this.north != comparisonGrid.north ||
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/analyst/Grid.java]: Yes, that's much cleaner.

File: src/main/java/com/conveyal/r5/analyst/TravelTimeComputer.java [hunk 4]
```diff
@@ -216,19 +225,90 @@ public void write (OutputStream os) throws IOException {
             // Create a new Raptor Worker.
             FastRaptorWorker worker = new FastRaptorWorker(network.transitLayer, request, accessTimes);
 
+            if (request.returnPaths || request.returnInVehicleTimes || request.returnWaitTimes) {
+                worker.saveAllStates = true; // By default, this is false and intermediate results (e.g. paths) are discarded.
+            }
+
             // Run the main RAPTOR algorithm to find paths and travel times to all stops in the network.
+
+            /* Total travel time from origin stop to a given stop in a given iteration of the algorithm  */
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/analyst/TravelTimeComputer.java]: Clarified the comment

File: src/main/java/com/conveyal/r5/analyst/TravelTimeComputer.java [hunk 5]
```diff
@@ -216,19 +225,90 @@ public void write (OutputStream os) throws IOException {
...
+                    }
+                }
+
+                perTargetPropagater = new PerTargetPropagater(egressModeLinkedDestinations, request, transitTravelTimesToStops, nonTransitTravelTimesToDestinations, inVehicleTimesToStops, waitTimesToStops, pathsToStops);
+                perTargetPropagater.reducer = new TravelTimeSurfaceReducer(request, network, os);
+                perTargetPropagater.pathWriter = new PathWriter(request, network, pathList);
+
+            } else {
+
+                perTargetPropagater = new PerTargetPropagater(egressModeLinkedDestinations, request, transitTravelTimesToStops, nonTransitTravelTimesToDestinations, null, null, null);
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/analyst/TravelTimeComputer.java]: Still a TODO

File: src/main/java/com/conveyal/r5/analyst/TravelTimeSurfaceReducer.java [hunk 6]
```diff
@@ -36,16 +38,18 @@
...
         this.network = network;
         this.outputStream = outputStream;
 
         try {
             // use an in-memory access grid, don't specify disk cache file
-            encodedResults = new AccessGridWriter(task.zoom, task.west, task.north, task.width, task.height, task.percentiles.length);
+            timeGrid = new TimeGrid(task.zoom, task.west, task.north, task.width, task.height, task.percentiles.length);
+            timeGrid.initialize("ACCESSGR", 0);
+
         } catch (IOException e) {
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/analyst/TravelTimeSurfaceReducer.java]: Still a TODO

File: src/main/java/com/conveyal/r5/analyst/cluster/PathWriter.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            }
+
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void finishPaths(){
+        try{
+            outputStream.close();
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/analyst/cluster/PathWriter.java]: Still a TODO

File: src/main/java/com/conveyal/r5/analyst/cluster/TimeGrid.java [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+ * Version, 4-byte integer
+ * (4 byte int) Web mercator zoom level
+ * (4 byte int) west (x) edge of the grid, i.e. how many pixels this grid is east of the left edge of the world
+ * (4 byte int) north (y) edge of the grid, i.e. how many pixels this grid is south of the top edge of the world
+ * (4 byte int) width of the grid in pixels
+ * (4 byte int) height of the grid in pixels
+ * (4 byte int) number of values per pixel
+ * (repeated 4-byte int) values of each pixel in row major order. Values are not delta coded here.
+ */
+public class TimeGrid {
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/analyst/cluster/TimeGrid.java]: Still a TODO

File: src/main/java/com/conveyal/r5/profile/PerTargetPropagater.java [hunk 9]
```diff
@@ -25,36 +26,55 @@
...
+    int [][] inVehicleTimesToStops;
+
+    /** Waiting time component of time from origin stop to a given stop in a given iteration of the algorithm  */
+    int [][] waitTimesToStops;
+
+    /** Index of path used from origin stop to a given stop in a given iteration of the algorithm  */
+    int [][] pathsToStops;
+
+
+    public PerTargetPropagater(LinkedPointSet targets, AnalysisTask task, int[][] travelTimesToStopsEachIteration, int[] nonTransitTravelTimesToTargets, int[][] inVehicleTimesToStops, int[][] waitTimesToStops, int[][] paths) {
```
[the remark this answers is not in the document]
Comment (ansoncfit) [reply src/main/java/com/conveyal/r5/profile/PerTargetPropagater.java]: Still a TODO

**Pull request merged**